packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data.ts

```diff
@@ -1,27 +1,35 @@
 import { useMemo } from "react";
 import { PropertyRow } from "./types";
-import { extractEnrichedPropertyTypesFromEntity } from "./extract-enriched-property-types-from-entity";
-import { sortRowData } from "../../../../../../../components/GlideGlid/utils";
+import { generatePropertyRowsFromEntity } from "./use-row-data/generate-property-rows-from-entity";
 import { useEntityEditor } from "../../entity-editor-context";
+import { fillRowDataIndentCalculations } from "./fill-row-data-indent-calculations";
+import { sortRowData } from "../../../../../../../components/GlideGlid/utils/sorting";
+import { flattenExpandedItemsOfTree } from "./flatten-expanded-items-of-tree";
 
 export const useRowData = () => {
-  const { rootEntityAndSubgraph, propertySort } = useEntityEditor();
+  const { rootEntityAndSubgraph, propertySort, propertyExpandStatus } =
+    useEntityEditor();
 
   const rowData = useMemo<PropertyRow[]>(() => {
     if (!rootEntityAndSubgraph) {
       return [];
     }
 
-    const enrichedPropertyTypes = extractEnrichedPropertyTypesFromEntity(
-      rootEntityAndSubgraph,
-    );
-
-    return enrichedPropertyTypes;
+    return generatePropertyRowsFromEntity(rootEntityAndSubgraph);
   }, [rootEntityAndSubgraph]);
 
-  const sortedRowData = useMemo(() => {
-    return sortRowData(rowData, propertySort);
-  }, [rowData, propertySort]);
+  const sortedRowData = useMemo(
+    () => sortRowData(rowData, propertySort),
+    [rowData, propertySort],
+  );
+
+  const flattenedRowData = flattenExpandedItemsOfTree(
+    sortedRowData,
+    propertyExpandStatus,
+    "rowId",
+  );
+
+  fillRowDataIndentCalculations(flattenedRowData);
```

Comment:
same https://github.com/hashintel/hash/pull/1266#discussion_r1011582582

Reply:
Sorry @yusufkinatas – I should have been clearer. You're mutating during React render, which breaks a rule of React. I'd encourage changing this to not mutate at all if possible, and especially not during React render as it must be side effect free. 

`fillRowDataIndentCalculations` mutates the `PropertyRow` objects generated by `generatePropertyRowsFromEntity`, which should probably be avoided anyway, and also you're flattening on every react render which can also be avoided. A patch like this should help. 

```diff
Index: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/fill-row-data-indent-calculations.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/fill-row-data-indent-calculations.ts b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/fill-row-data-indent-calculations.ts
--- a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/fill-row-data-indent-calculations.ts	(revision 0d27534690cd698bfff0a1e13cb883f92c0d9afb)
+++ b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/fill-row-data-indent-calculations.ts	(date 1667408366266)
@@ -42,8 +42,13 @@
       }
     }
 
-    property.verticalLinesForEachIndent = arr;
+    rowData[row] = {
+      ...property,
+      verticalLinesForEachIndent: arr,
+    };
   }
+
+  return rowData;
 };
 
 /**
@@ -103,6 +108,8 @@
       }
     }
   }
+
+  return rowData;
 };
 
 /**
@@ -111,8 +118,9 @@
  * @param rowData is the flattened tree of property rows
  */
 export const fillRowDataIndentCalculations = (rowData: PropertyRow[]) => {
-  markUpperHalvesOfIndentationLines(rowData);
-  markBottomHalvesOfIndentationLines(rowData);
+  return markBottomHalvesOfIndentationLines(
+    markUpperHalvesOfIndentationLines([...rowData]),
+  );
 };
 
 /**
Index: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data.ts b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data.ts
--- a/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data.ts	(revision 0d27534690cd698bfff0a1e13cb883f92c0d9afb)
+++ b/packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/use-row-data.ts	(date 1667408365223)
@@ -23,12 +23,14 @@
     [rowData, propertySort],
   );
 
-  const flattenedRowData = flattenExpandedItemsOfTree(
-    sortedRowData,
-    propertyExpandStatus,
-  );
+  const flattenedRowData = useMemo(() => {
+    const flattenedRowData = flattenExpandedItemsOfTree(
+      sortedRowData,
+      propertyExpandStatus,
+    );
 
-  fillRowDataIndentCalculations(flattenedRowData);
+    return fillRowDataIndentCalculations(flattenedRowData);
+  }, [sortedRowData, propertyExpandStatus]);
 
   return flattenedRowData;
 };
```

Reply:
Thinking about this, do you have a problem that now `verticalLinesForEachIndent` is only set on the `PropertyRow` at the top level (which has been flattened), not on any children?

Reply:
Let me know if you want to pair on this.

Reply:
Let me know if you want to pair on this.

Reply:
Hey Nate, have a look at this commit https://github.com/hashintel/hash/pull/1266/commits/c3e7d60fb20318261606837cf4576e827f8a6c6b

I feel like this should be enough for this PR, and and if we decide to do further changes (e.g. holding lines on a separate state OR changing the line calculation logic to just hold the lines as a whole instead of per-cell) we should be doing that in a follow-up PR.

But with the commit above, `flattenExpandedItemsOfTree` and `fillRowDataIndentCalculations` only runs when `sortedRowData` or `propertyExpandStatus` changes, with is intended.